### ISSUE TITLE
What is the purpose of parameter `applyStack` of `Doc`?

### ISSUE THREAD

**Issue by nieyuyao** [share/sharedb#474]
I'm reading the code for this `lib/client/doc.js`. I don't understand the purpose of parameter `applyStack`. Especially in the next code, the for loop seems useless. Parameter `stackLength ` is equal to `this.applyStack.length`.

``` javascript
if (!source && this.type === types.defaultType && op.op.length > 1) {
      if (!this.applyStack) this.applyStack = [];
      var stackLength = this.applyStack.length;
      for (var i = 0; i < op.op.length; i++) {
        var component = op.op[i];
        var componentOp = {op: [component]};
        // Transform componentOp against any ops that have been submitted
        // sychronously inside of an op event handler since we began apply of
        // our operation
        for (var j = stackLength; j < this.applyStack.length; j++) {
          var transformErr = transformX(this.applyStack[j], componentOp);
          if (transformErr) return this._hardRollback(transformErr);
        }
        // Apply the individual op component
        this.emit('before op', componentOp.op, source, op.src);
        this.data = this.type.apply(this.data, componentOp.op);
        this.emit('op', componentOp.op, source, op.src);
      }
      this.emit('op batch', op.op, source);
      // Pop whatever was submitted since we started applying this op
      this._popApplyStack(stackLength);
      return;
    }
```
I look forward to your answers.

**Comment by happy15** [comment]
moved from #475:
1. what's the meaning of "applying an op incrementally"?
2. is it possible for applyStack.length to get longer during _otApply?

**Comment by alecgibson** [comment]
# What is "applying an op incrementally"?

`json0` ops can be "shattered" and applied incrementally. That is, a `json0` op might look like:

```js
[
  {p: ['a'], oi: 'foo'},
  {p: ['b'], oi: 'bar'},
]
```

Shattering this op and applying it incrementally means you'll get two `'op'` events from **remote** ops (and just one `'op batch'` event):

Event 1:

```js
[{p: ['a'], oi: 'foo'}]
```

Event 2:

```js
[{p: ['b'], oi: 'bar'}]
```

# Is it possible for `applyStack.length` to get longer?

Yes. If you submit an op inside an `'op'` event, you will push more ops onto the apply stack before it's been cleared.

**Comment by nieyuyao** [comment]
OK, I get it a little bit. We can't use this parameter at most of the time?

**Comment by alecgibson** [comment]
I'm not sure what you're trying to achieve?

**Comment by happy15** [comment]
> ## What is "applying an op incrementally"?
> `json0` ops can be "shattered" and applied incrementally. That is, a `json0` op might look like:
> 
> ```js
> [
>   {p: ['a'], oi: 'foo'},
>   {p: ['b'], oi: 'bar'},
> ]
> ```
> 
> Shattering this op and applying it incrementally means you'll get two `'op'` events from **remote** ops (and just one `'op batch'` event):
> 
> Event 1:
> 
> ```js
> [{p: ['a'], oi: 'foo'}]
> ```
> 
> Event 2:
> 
> ```js
> [{p: ['b'], oi: 'bar'}]
> ```
> 
> ## Is it possible for `applyStack.length` to get longer?
> Yes. If you submit an op inside an `'op'` event, you will push more ops onto the apply stack before it's been cleared.

so applyStack is only used for **remote** op and not for local op, am I right?

**Comment by alecgibson** [comment]
> so applyStack is only used for remote op and not for local op, am I right?

Um, you're not really right. The point of the `applyStack` is to catch local ops that were submitted whilst applying a shattered remote op.

So if we get a remote op with multiple components, we [create the `applyStack`](https://github.com/share/sharedb/blob/326c5abf03faa32e8f1e59cdee4666dc65089f53/lib/client/doc.js#L621-L622).

Then, if the local client submits an op while we're applying (usually in response to the synchronous [`'op'` hook](https://github.com/share/sharedb/blob/326c5abf03faa32e8f1e59cdee4666dc65089f53/lib/client/doc.js#L637)), then that **local** op is [pushed onto the `applyStack`](https://github.com/share/sharedb/blob/326c5abf03faa32e8f1e59cdee4666dc65089f53/lib/client/doc.js#L779-L783), because `_pushOp` is called [when submitting](https://github.com/share/sharedb/blob/326c5abf03faa32e8f1e59cdee4666dc65089f53/lib/client/doc.js#L763).

**Comment by happy15** [comment]
> > so applyStack is only used for remote op and not for local op, am I right?
> 
> Um, you're not really right. The point of the `applyStack` is to catch local ops that were submitted whilst applying a shattered remote op.
> 
> So if we get a remote op with multiple components, we [create the `applyStack`](https://github.com/share/sharedb/blob/326c5abf03faa32e8f1e59cdee4666dc65089f53/lib/client/doc.js#L621-L622).
> 
> Then, if the local client submits an op while we're applying (usually in response to the synchronous [`'op'` hook](https://github.com/share/sharedb/blob/326c5abf03faa32e8f1e59cdee4666dc65089f53/lib/client/doc.js#L637)), then that **local** op is [pushed onto the `applyStack`](https://github.com/share/sharedb/blob/326c5abf03faa32e8f1e59cdee4666dc65089f53/lib/client/doc.js#L779-L783), because `_pushOp` is called [when submitting](https://github.com/share/sharedb/blob/326c5abf03faa32e8f1e59cdee4666dc65089f53/lib/client/doc.js#L763).

thanks, I kinda get you. 

but what if local ops are submitted in 'before op batch' or 'before op' hook? 

say remote op = [cp1, cp2, cp3] (cp stands for component):
1. after arriving at local client, [cp1, cp2, cp3] are immediately transformed against pendingOps at local before _otApply and are fit for the state of local data, say data1. 
2. right after entering _otApply, 'before op batch' is fired and local ops = [op1, op2, op3, op4] are submitted and applied to  local data causing the state of local data to become data2.
3. now it starts to apply cp1. applyStack is initialized as [], so there is nothing to be transformed against for cp1.
4. cp1 is applied to the local data which is data2 not data1.

is the process above possible?

**Comment by alecgibson** [comment]
I think that sounds about right.

**Comment by happy15** [comment]
> I think that sounds about right.

is it a problem?

**Comment by alecgibson** [comment]
I don't know — you tell me! Have you found a bug?

**Comment by happy15** [comment]
> > > so applyStack is only used for remote op and not for local op, am I right?
> > 
> > 
> > Um, you're not really right. The point of the `applyStack` is to catch local ops that were submitted whilst applying a shattered remote op.
> > So if we get a remote op with multiple components, we [create the `applyStack`](https://github.com/share/sharedb/blob/326c5abf03faa32e8f1e59cdee4666dc65089f53/lib/client/doc.js#L621-L622).
> > Then, if the local client submits an op while we're applying (usually in response to the synchronous [`'op'` hook](https://github.com/share/sharedb/blob/326c5abf03faa32e8f1e59cdee4666dc65089f53/lib/client/doc.js#L637)), then that **local** op is [pushed onto the `applyStack`](https://github.com/share/sharedb/blob/326c5abf03faa32e8f1e59cdee4666dc65089f53/lib/client/doc.js#L779-L783), because `_pushOp` is called [when submitting](https://github.com/share/sharedb/blob/326c5abf03faa32e8f1e59cdee4666dc65089f53/lib/client/doc.js#L763).
> 
> thanks, I kinda get you.
> 
> but what if local ops are submitted in 'before op batch' or 'before op' hook?
> 
> say remote op = [cp1, cp2, cp3] (cp stands for component):
> 
> 1. after arriving at local client, [cp1, cp2, cp3] are immediately transformed against pendingOps at local before _otApply and are fit for the state of local data, say data1.
> 2. right after entering _otApply, 'before op batch' is fired and local ops = [op1, op2, op3, op4] are submitted and applied to  local data causing the state of local data to become data2.
> 3. now it starts to apply cp1. applyStack is initialized as [], so there is nothing to be transformed against for cp1.
> 4. cp1 is applied to the local data which is data2 not data1. since cp1 is not transformed against [op1, op2, op3, op4], cp1 is not fit for data2.
> 
> is the process above possible?

I tried to make a test case which demonstrates the case described:
```
  describe('applyStack', function() {
    ...
    it('remote ops are transformed by ops submitted in `before op` event handlers', function(done) {
      var doc = this.doc
      var doc2 = this.doc2;
      var doc3 = this.doc3;
      doc2.submitOp({p:['list'], oi:[]}, () => {
        doc.fetch(() => {
          // console.log('doc.data', doc.data)
          // console.log('doc2.data', doc2.data)

          var opFromBeforeOpEvent = 1
          doc.on('before op', (op, source) => {
            console.log('\nbefore op @ doc', op, source ? 'local' : 'remote')
            if (source) {
              return
            }
            if (opFromBeforeOpEvent <= 0) {
              return
            }
            opFromBeforeOpEvent--
            // console.log('okkkkkk')
            doc.submitOp({p:['list', 0], li: 2}, {source: true})
          })

          var opFromOpEvent = 1
          doc.on('op', (op, source) => {
            console.log('\nop @ doc', op, source ? 'local' : 'remote')
            if (source) {
              return
            }
            if (opFromOpEvent <= 0) {
              return
            }
            opFromOpEvent--
            doc.submitOp({p:['list', 0], li: 3}, {source: true})
          })
          doc2.submitOp([{p: ['list', 0], li: 1}, {p: ['list', 1], li: 42}], () => {
            doc.fetch()
            verifyConsistency(doc, doc2, doc3, [], done)
          })
        })
      })
    })
    ...
  }
```
and it's failing
```
      Uncaught AssertionError: expected { list: [ 3, 1, 2, 42 ] } to deeply equal { list: [ 3, 2, 1, 42 ] }
      + expected - actual

       {
         "list": [
           3
      +    2
           1
      -    2
           42
         ]
       }
```
logs:
```
doc.data { list: [ 3, 1, 2, 42 ] }
doc2.data { list: [ 3, 2, 1, 42 ] }
doc3.data { list: [ 3, 2, 1, 42 ] }
```

If what I made is correct, then a fix could be move emit of 'before op' event before the transform against applyStack:
```
       for (var i = 0; i < op.op.length; i++) {
         var component = op.op[i];
         var componentOp = {op: [component]};
+        // Apply the individual op component
+        this.emit('before op', componentOp.op, source, op.src);
+        console.log(`applyStack`, this.name, this.applyStack.map(x => x.op))
         // Transform componentOp against any ops that have been submitted
         // sychronously inside of an op event handler since we began apply of
         // our operation
@@ -631,8 +635,6 @@ Doc.prototype._otApply = function(op, source) {
           var transformErr = transformX(this.applyStack[j], componentOp);
           if (transformErr) return this._hardRollback(transformErr);
         }
-        // Apply the individual op component
-        this.emit('before op', componentOp.op, source, op.src);
         this._setData(this.type.apply(this.data, componentOp.op));
         this.emit('op', componentOp.op, source, op.src);
       }
```
and it will pass along with other 625 cases.
```
626 passing (3s)
```

Again, thanks for the awesome lib and what I am trying to do is just to understand it. Respect to all the hard work you guys have done.

**Comment by alecgibson** [comment]
Okay good catch! So if I understand correctly, this is the sequence of events for your test case:

1. We create an empty list and get our docs up-to-date
2. `doc2` submits an op with two `li` components, resulting in the local array `[1, 42]`
3. `doc` receives the remote op components. We start iterating over them.
4. `doc`'s `before op` hook is triggered by the first component. Its `data` is still an empty array `[]`
5. `doc` submits an op in this `before op` hook. Since an `applyStack` has been created, (because of the incoming remote op), the local op is pushed onto the `applyStack`.
6. `doc` applies its own op, and now has data `[2]`
7. `doc` applies the first op component `p: [0], li: 1` — **without transforming** —, and gets `[1, 2]`
8. `doc`'s `op` hook is triggered, and it submits _another_ op, which is _also_ pushed onto the `applyStack`. It results in `[3, 1, 2]`
9. We finally apply the second remote op component to `doc`. We now have both local ops in the `applyStack`, which transforms the remote op component from `p: [1], li: 42` to `p: [3], li: 42`, and results in the array `[3, 1, 2, 42]` when applied.
10. In contrast, `doc2` has applied its ops locally already, and has `[1, 42]`. It receives the first of `doc`'s ops (from `before op`): `p: 0, li: 2`, and applies it: `[2, 1, 42]`
11. It then receives the second op (submitted in `op`): `p: 0, li: 3` and applies it: `[3, 2, 1, 42]`, which **doesn't match** `doc`'s data.

The fix you've proposed is to move the `before op` event **before** the `applyStack` transformation loop. The result of  this is that any ops submitted in the `before op` hook are included in the transformation of the current op component being applied.

I think this is the right thing to do. You've mutated the local data by applying an op, so it makes sense to transform the incoming op by this mutation so that it's properly compatible. As a sense check, it also moves the `before op` event close to the `before op batch` event, which is probably also more logical, and which is why this bug wouldn't happen if submitting an op in the `before op batch` event, which also precedes the `applyStack` transformation loop.

Trying to think about some potential gotchas:

 - moving the `before op` event above the transformation means you may get a transformation error (and subsequent hard rollback) between `before op` and `op`. I'm not sure this is a particularly bad case, since we do a hard rollback anyway.
 - we're slightly semantically changing `before op` — at the moment it sort of means "right before an op is about to be applied", but we'd be redefining it to something a bit looser, like "an op will be applied very soon, but we're going to do a bit more work first". Again, this is nitpicky, and I'm not entirely sure it's necessarily a cause for concern.

In summary, I think this is a good bug catch with a good fix — could you please raise a Pull Request?
